Fix bug #863

zagge-cgeo opened the pull request:
This patch fixes the bug #863

Commit:
Fix bug #863

jreiser commented:
This patched line has problems:
`    if (upx_uint64_t(edir.functions + edir.names) * 4 >= upx_uint64_t(esize)) {`
Because both `edir:functions` and `edir:names` are (32-bit) `unsigned`, then so is their sum, and wrap-around (overflow) is ignored.  Creating a 64-bit value from the 32-bit sum does not recover the overflow.  Instead, one of the operands must be converted to 64-bit before the addition.

Also, what is the magic number `4`?  It should be `sizeof` something, or a named `const` with explanatory comment.

Finally, the test is somewhat weak.  In conf.h: `#define UPX_RSIZE_MAX_MEM (768 * 1024 * 1024) // DO NOT CHANGE !!!`, and the implied space associated with each of `functions`, `names`, and `esize` must be less, including any initial offset and trailing suffix in the file.  Figure out if comparing one of them to `UPX_RSIZE_MAX_MEM` would be better; or at least make a comment that the code isn't checking that constraint.

The current tip of branch `devel` d00ba4b4656837c08841bd8f9599fda87bb7c7d2 has different code than shown in the patch.  Please update.

Commit:
Improve patch

Implement things pointed out by jreiser on review of the patch.
- Fix type conversion issue
- Added constant with comment
- Added comment for not validating UPX_RSIZE_MAX_MEM

Additional changes
- Honor the size of the export_dir_t when calculating the available
  space

zagge-cgeo commented:
Can this please be merged or is there something for me to update t get it merged?

jreiser commented:
Subsumed by developer commit in #863.
